### PR TITLE
issue/2747 Added questionModel attemptStates API

### PR DESCRIPTION
**Pull request by olivermartinfoster**
#2747 

To keep the previous question attempt states for use in plugins.

Requires [#192](https://github.com/adaptlearning/adapt-contrib-spoor/pull/192) spoor pr to be saved and restored across sessions.

#### Added
* `trackableType` property throughout to list the type of the `trackable` attributes
* `questionModel.getAttemptState` Returns the raw state of the component in a format for spoor to store
* `questionModel.getAttemptObject` Returns a formatted state object which can be passed through to `questionModel.set`
* `questionModel.setAttemptObject` Sets the current state of the component from a state object
* `questionModel.addAttemptObject` Adds an attempt object to the attemptStates store
* `questionModel.getAttemptObjects` Returns an array of the component attempt state objects
* `_spoor._tracking._shouldStoreAttempts` to `config.json`
* Question completion automatically adds a state object to the attempt state store

#### Testing
* Add component 'c-60' mcq `_isResetOnRevisit: "soft"` 
* Keep attempting 'c-60'
* Use the console to determine mcq attempts
```js
var Adapt = require('core/js/adapt');
Adapt.findById('c-60').getAttemptObjects();
```
* To see states restored across sessions use spoor pr [#192](https://github.com/adaptlearning/adapt-contrib-spoor/pull/192)
* And enable `_spoor._tracking._shouldStoreResponses` and `_spoor._tracking._shouldStoreAttempts` in `config.json`

### CODE REVIEW

File: src/core/js/models/questionModel.js [hunk 1]
```diff
@@ -31,8 +31,30 @@ define([
         '_isSubmitted',
         '_score',
         '_isCorrect',
-        '_attemptsLeft'
+        '_attemptsLeft',
+        '_attemptStates'
       ]);
+    },
```
Comment (tomgreenfield) [inline src/core/js/models/questionModel.js:37]: Cannot compile this branch due to commas here and on line 47.

File: src/core/js/models/questionModel.js [hunk 2]
```diff
@@ -31,8 +31,30 @@ define([
...
+        Boolean,
+        Number,
+        Array
+      ]);
+    },
+
+    setupModel(...args) {
+      super.setupModel(...args);
+      this.on('change:_isInteractionComplete', (model, value) => {
+        if (!value) {
```
Comment (tomgreenfield) [inline src/core/js/models/questionModel.js:52]: For clarity, suggest:
```suggestion
      this.on('change:_isInteractionComplete', (model, isInteractionComplete) => {
        if (!isInteractionComplete) {
```

File: src/core/js/models/questionModel.js [hunk 3]
```diff
@@ -320,6 +342,98 @@ define([
...
+     * Returns the current attempt state raw data or the raw data from the supplied attempt state object.
+     * @param {Object} [object] JSON object representing the component state. Defaults to current JSON.
+     * @returns {Array}
+     */
+    getAttemptState(object = this.toJSON()) {
+      const trackables = this.trackable();
+      const types = this.trackableType();
+      trackables.find((name, index) => {
+        if (name !== '_attemptStates') return;
+        // Exclude _attemptStates as it's trackable but isn't needed here
```
Comment (tomgreenfield) [inline src/core/js/models/questionModel.js:355]: Please move comment to sit before the conditional (see also line 381).

File: src/core/js/models/questionModel.js [hunk 4]
```diff
@@ -320,6 +342,98 @@ define([
...
+
+    /**
+     * Returns an array of the previous state objects. The most recent state is last in the list.
+     * @returns {Array}
+     */
+    getAttemptObjects() {
+      const states = this.get('_attemptStates') || [];
+      return states.map(state => {
+        return this.getAttemptObject(state);
+      })
```
Comment (tomgreenfield) [inline src/core/js/models/questionModel.js:434]:
```suggestion
      return states.map(state => this.getAttemptObject(state));
```

**Comment by oliverfoster** [comment]
I realised that the `_userAttempts` wasn't getting stored because the view is attached to `change:_isSubmitted` after the model and so the model would have to wait for the view to finish. I've instead added a `QuestionModel.prototype.onSubmitted` function to complement the `QuestionView.prototype.onSubmitted` function called at the end of `QuestionView.prototype.onSubmitClicked`.

**Comment by oliverfoster** [comment]
@tomgreenfield if you could re-review this one is ready to go in

**Pull request merged**